File: LongestIncreasingSubsequence.java
import java.util.*;

public class LongestIncreasingSubsequence {
  public int lengthOfLIS(int[] nums) {
      int[] dp = new int[nums.length];
      Arrays.fill(dp, 1);
     int finalMax = 1;
      for (int i = nums.length - 1; i >= 0; i--) {
        for (int j = i; j < nums.length; j++) {
          if (nums[j] > nums[i]) {
            dp[i] = Math.max(dp[j] + 1, dp[i]);
            finalMax = Math.max(finalMax, dp[i]);
          }
        }
      }
      return finalMax;
  }
}

File: LongestIncreasingSubsequence.py
class LongestIncreasingSubsequence(object):
    def lengthOfLIS(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        
        dp = []
        for i in range(len(nums)):
            dp.append(1)
        maxLength = 1
        
        for i in reversed(range(len(nums))):
            for j in range (i + 1, len(nums)):
                if nums[i] < nums[j]:
                    dp[i] = max(dp[j] + 1, dp[i])
                    maxLength = max(dp[i], maxLength)
                
        return maxLength
                